index/types.go

```diff
@@ -21,22 +21,96 @@
 package index
 
 import (
-	"github.com/m3db/m3ninx/index/segment"
+	"regexp"
 
-	"github.com/m3db/m3x/instrument"
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/postings"
+
+	"github.com/m3db/m3x/errors"
 )
 
-// Index is a collection of segments.
+// Index is a collection of searchable documents.
 type Index interface {
-	segment.Readable
-	segment.Writable
+	// Insert inserts the given document into the index. The document is guaranteed to be
+	// searchable once the Insert method returns.
+	Insert(d doc.Document) error
+
+	// Snapshot returns a point-in-time snapshot of the index that provides a consistent
+	// view for searches.
+	Snapshot() (Snapshot, error)
+
+	// Seal marks the index as immutable. After Seal is called no more documents can be
+	// inserted into the index.
+	Seal() error
+
+	// Close closes the index and releases any internal resources.
+	Close() error
+}
+
+// Snapshot is a point-in-time snapshot of an index. It must be closed to
+// release its internal readers.
+type Snapshot interface {
+	// Readers returns the set of readers contained in the snapshot. Each call to Readers
+	// will return independent clones of the same Readers which the caller is responsible
+	// for closing.
+	Readers() Readers
+
+	// Close closes the snapshot and releases any internal resources.
+	Close() error
+}
+
+// Reader provides a point-in-time accessor to the documents in an index.
+type Reader interface {
+	// MatchTerm returns a postings list over all documents which match the given term.
+	MatchTerm(field, term []byte) (postings.List, error)
+
+	// MatchRegexp returns a postings list over all documents which match the given
+	// regular expression.
+	MatchRegexp(field, regexp []byte, compiled *regexp.Regexp) (postings.List, error)
+
+	// Docs returns an iterator over the documents whose IDs are in the provided
+	//  postings list.
+	Docs(pl postings.List) (doc.Iterator, error)
+
+	// Clone returns an independent clone of this Reader. The returned reader must be
+	// closed independently of this Reader.
+	Clone() (Reader, error)
+
+	// Close closes the reader and releases any internal resources.
+	Close() error
+}
+
+// Readers is a slice of Reader.
+type Readers []Reader
+
+// Close closes all of the Readers in rs.
+func (rs Readers) Close() error {
+	multiErr := errors.NewMultiError()
+	for _, r := range rs {
+		err := r.Close()
+		if err != nil {
+			multiErr.Add(err)
+		}
+	}
+
+	if multiErr.Empty() {
+		return nil
+	}
+	return multiErr
 }
 
-// Options is a set of knobs by which to tweak Index-ing behaviour.
-type Options interface {
-	// SetInstrumentOptions sets the instrument options.
-	SetInstrumentOptions(value instrument.Options) Options
+// Clone returns a clone of all of the internal Readers.
+func (rs Readers) Clone() (Readers, error) {
```

Comment:
Where is this used?

Reply:
Talked offline, this is used in the `search` package when we want to have multiple independent versions of the same Reader.